pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java

```diff
@@ -0,0 +1,353 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.scheduling.support.CronExpression;
+import org.springframework.stereotype.Component;
+
+/**
+ * This class is responsible for purging the DWH snapshots when they expire. It also exposes few
+ * helper methods to access the DWH files and directories.
+ */
+@EnableScheduling
+@Component
+public class DWHFilesManager {
+
+  @Autowired private DataProperties dataProperties;
+
+  private CronExpression purgeCron;
+
+  private LocalDateTime lastPurgeRunEnd;
+
+  private volatile boolean isPurgeJobRunning;
+
+  private static final Logger logger = LoggerFactory.getLogger(DWHFilesManager.class.getName());
+
+  @PostConstruct
+  public void init() {
+    purgeCron = CronExpression.parse(dataProperties.getPurgeSchedule());
+  }
+
+  private void setLastPurgeRunStatus(PurgeRunStatus status) {
+    if (status == PurgeRunStatus.SUCCESS) {
+      lastPurgeRunEnd = LocalDateTime.now();
+    }
+  }
+
+  private boolean isPurgeJobRunning() {
+    return isPurgeJobRunning;
+  }
+
+  private void setPurgeJobRunning(boolean isPurgeJobRunning) {
+    this.isPurgeJobRunning = isPurgeJobRunning;
+  }
+
+  private synchronized boolean lockPurgeJob() {
+    if (isPurgeJobRunning()) {
+      // Already the job is running so lock can't be acquired.
+      return false;
+    }
+    setPurgeJobRunning(true);
+    return true;
+  }
+
+  private synchronized void releasePurgeJob() {
+    setPurgeJobRunning(false);
+  }
+
+  /**
+   * @return the next scheduled time to run the purge job based on the previous run time.
+   */
+  LocalDateTime getNextPurgeTime() {
+    if (lastPurgeRunEnd == null) {
+      return LocalDateTime.now();
+    }
+    return purgeCron.next(lastPurgeRunEnd);
+  }
+
+  // Every 30 seconds, check if the purge job needs to be triggered.
+  @Scheduled(fixedDelay = 30000)
+  private void checkPurgeSchedule() {
+    try {
+      if (!lockPurgeJob()) {
+        return;
+      }
+      LocalDateTime next = getNextPurgeTime();
+      logger.info("Last purge run was at {} next run is at {}", lastPurgeRunEnd, next);
+      if (next.compareTo(LocalDateTime.now()) <= 0) {
+        logger.info("Purge run triggered at {}", LocalDateTime.now());
+        purgeDWHFiles();
+        logger.info("Purge run completed at {}", LocalDateTime.now());
+      }
+    } finally {
+      releasePurgeJob();
+    }
+  }
+
+  /**
+   * This method will scan the DWH base directory to check for any older DWH snapshots that needs to
+   * be purged. It will retain the most recent successfully completed snapshots and purge the older
+   * ones. Any incomplete snapshots will not be purged by this job and will have to be manually
+   * verified and acted upon.
+   */
+  private void purgeDWHFiles() {
```

Comment:
Question: Have you looked into how these methods can be unit-tested, e.g., using an in memory file-system? Our file handling logic in the controller is becoming more complex and I am wondering how hard it is to start writing unit-tests for these.

Reply:
I have added a few unit tests for the LocalFileSystem, this is written by manually handling the creation and deletion of  directories + files. Will explore on the in-memory file system and get back.

Reply:
Okay, thanks for adding the new unit-tests. Please feel free to resolve this conversation if adding tests for the GCS case is too much work.

Reply:
Have added few test cases for GCS case as well, resolving this for now.